chapter_01/final_exercise_01.md

```diff
@@ -20,27 +19,142 @@
+Before hadoop, storage and processing of large quantaties of data was very hard and required a lot of expensive hardware and cost a lot to run.
+after hadoop came, it allowed to store large quantaties of data and to scale to an unlimited number of nodes which was not possible before.
+
+Hadoop started in 2006 by Doug Cutting (working at yahoo at the time) after he and Mike Cafarella had worked on the apache Nutch project which was supposed to be a search engine which could index 1 billion pages but after Doug realized that it could only scale to 20-40 nodes and that it would be too expensive to run, so Doug decided to create a new project named hadoop (the name of his son's yellow elephant toy) and in 2007, yahoo was able to scale hadoop to 1000 nodes and use it,
+in 2008, yahoo has open sourced the project and in july they were able to scale it to 4000 nodes.
+In 2009, Doug left Yahoo and went to work at cloudera, in 2011 Apache released Apache hadoop version 1.0 and in 2013 Hadoop 2.0.6 was available and in 2017 version 3.0 came out.
+The main problems with hadoop are that hadoop is very bad with small files because of the way it stores data, for each block, the namenode stores metadata about the files (the locations of them in datanodes and the file system structure).
+now, instead of managing a small number of computers, we need to manage all the nodes either ourselves or through a vendor providing us managed hadoop cluster.
+
+
+
 2. **Q:** Can you name a few vendor-specific distributions of Hadoop?
+2. **A:** *Cloudera Distribution Platform* - A managed distribution by cluodera which includes a cloudera cluster manager, integrated security and data governance, *Amazon Elastic MapReduce* - A managed distribution of hadoop with a pay-per-use billing system, automatic scaling and easy integration with other aws products, *Microsoft azure HDInsight* - includes integration with microsoft tools like Power BI and Azure data factory, also billed as a pay-per-use patform.
+
+
+
 3. **Q:** What are the three modes in which Hadoop can run?
+3. **A:** **Standalone (default)** - where it runs as a single java process, used to debug and make simple POC's just to test on a local machine with a small amount of data and is easy and quick to setup, mostly used for quick tests, **Psuedo Distributed** - where there is a single node cluster (many processes on one machine) which simulates the real world but still runs on one machine which is used to test and debug more like a staging area but is more resource intensive and takes more time to setup, **distributed** - a multi node cluster which includes the Namenode, Secondary Namenode and the Datanodes, this is used ion the real world where we need to actually process and store very large amounts of data quickly and efficiently, it is what we use to make big data queries and get valuable knowledge from them.
```

Comment:
Regarding the pseudu distributed mode. Are my processes running on one JVM or multiple?